### PR TITLE
[Android] Implement check in delegate

### LINKED ISSUE #32536: [Android] Implement Android Check-In Delegate
I knew I had to implement it according to your opinion. However, because I had to make sure that I could receive the check-in message as the first step, I implemented it using the Default class.

_Originally posted by @joonhaengHeo in https://github.com/project-chip/connectedhomeip/pull/32476#discussion_r1517206768_

### PR DESCRIPTION
**Pull request by joonhaengHeo**
Fix #32536

### CODE REVIEW

File: src/controller/java/AndroidCheckInDelegate.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    RefreshKeySender * OnKeyRefreshNeeded(ICDClientInfo & clientInfo, ICDClientStorage * clientStorage) override;
+    void OnKeyRefreshDone(RefreshKeySender * refreshKeySender, CHIP_ERROR error) override;
+
+    CHIP_ERROR SetDelegate(jobject checkInDelegateObj);
+
+private:
+    ICDClientStorage * mpStorage        = nullptr;
+    InteractionModelEngine * mpImEngine = nullptr;
+
+    chip::JniGlobalReference mCheckInDelegate;
```
Comment (yunhanw-google) [inline src/controller/java/AndroidCheckInDelegate.h:48]: no need for chip:: ?

Comment (joonhaengHeo) [reply src/controller/java/AndroidCheckInDelegate.h]: modified. Thanks.

File: src/controller/java/src/chip/devicecontroller/ICDCheckInDelegate.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+   */
+  @Nullable
+  byte[] onKeyRefreshNeeded(@Nonnull ICDClientInfo info);
+
+  /**
+   * Callback used to let the application know that the re-registration process is done.
+   *
+   * @param errorCode to check for success and failure
+   */
+  void onKeyRefreshDone(int errorCode);
```
Comment (yunhanw-google) [inline src/controller/java/src/chip/devicecontroller/ICDCheckInDelegate.java:54]: the error code should be long?

File: src/controller/java/AndroidCheckInDelegate.cpp [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+RefreshKeySender * AndroidCheckInDelegate::OnKeyRefreshNeeded(ICDClientInfo & clientInfo, ICDClientStorage * clientStorage)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    RefreshKeySender::RefreshKeyBuffer newKey;
+
+    bool isSetKey = false;
+    if (mCheckInDelegate.HasValidObjectRef())
+    {
+        JNIEnv * env = chip::JniReferences::GetInstance().GetEnvForCurrentThread();
```
Comment (yunhanw-google) [inline src/controller/java/AndroidCheckInDelegate.cpp:83]: check env is nullptr

File: src/controller/java/AndroidCheckInDelegate.cpp [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        VerifyOrReturnValue(err == CHIP_NO_ERROR, nullptr,
+                            ChipLogProgress(ICD, "onKeyRefreshNeeded - FindMethod is failed! : %" CHIP_ERROR_FORMAT, err.Format()));
+
+        jbyteArray key = static_cast<jbyteArray>(env->CallObjectMethod(mCheckInDelegate.ObjectRef(), onKeyRefreshNeededMethodID,
+                                                                       peerNodeId, startCounter, offset, monitoredSubject,
+                                                                       jniICDAesKey.jniValue(), jniICDHmacKey.jniValue()));
+
+        if (key != nullptr)
+        {
+            chip::JniByteArray jniKey(env, key);
```
Comment (yunhanw-google) [inline src/controller/java/AndroidCheckInDelegate.cpp:99]: chip:: is not necessary in this file?

File: src/controller/java/AndroidCheckInDelegate.cpp [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    env->CallVoidMethod(mCheckInDelegate.ObjectRef(), onCheckInCompleteMethodID, peerNodeId, startCounter, offset, monitoredSubject,
+                        jniICDAesKey.jniValue(), jniICDHmacKey.jniValue());
+}
+
+RefreshKeySender * AndroidCheckInDelegate::OnKeyRefreshNeeded(ICDClientInfo & clientInfo, ICDClientStorage * clientStorage)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    RefreshKeySender::RefreshKeyBuffer newKey;
+
+    bool isSetKey = false;
```
Comment (yunhanw-google) [inline src/controller/java/AndroidCheckInDelegate.cpp:80]: hasSetKey?

File: src/controller/java/AndroidCheckInDelegate.cpp [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+void AndroidCheckInDelegate::OnCheckInComplete(const ICDClientInfo & clientInfo)
+{
+    ChipLogProgress(
+        ICD, "Check In Message processing complete: start_counter=%" PRIu32 " offset=%" PRIu32 " nodeid=" ChipLogFormatScopedNodeId,
+        clientInfo.start_icd_counter, clientInfo.offset, ChipLogValueScopedNodeId(clientInfo.peer_node));
+
+    VerifyOrReturn(mCheckInDelegate.HasValidObjectRef(), ChipLogProgress(ICD, "check-in delegate is not implemented!"));
+
+    JNIEnv * env = chip::JniReferences::GetInstance().GetEnvForCurrentThread();
```
Comment (yunhanw-google) [inline src/controller/java/AndroidCheckInDelegate.cpp:62]: check whether env is nullptr

File: src/controller/java/AndroidCheckInDelegate.cpp [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    RefreshKeySender::RefreshKeyBuffer newKey;
+
+    bool isSetKey = false;
+    if (mCheckInDelegate.HasValidObjectRef())
+    {
+        JNIEnv * env = chip::JniReferences::GetInstance().GetEnvForCurrentThread();
+
+        PARSE_CLIENT_INFO(clientInfo, peerNodeId, startCounter, offset, monitoredSubject, jniICDAesKey, jniICDHmacKey)
+
+        jmethodID onKeyRefreshNeededMethodID;
```
Comment (yunhanw-google) [inline src/controller/java/AndroidCheckInDelegate.cpp:87]: jmethodID onKeyRefreshNeededMethodID = nullptr

File: src/controller/java/src/chip/devicecontroller/ICDCheckInDelegate.java (hunk 2 above)
Comment (joonhaengHeo) [reply src/controller/java/src/chip/devicecontroller/ICDCheckInDelegate.java:54]: Modified. Thanks.

File: src/controller/java/AndroidCheckInDelegate.cpp (hunk 3 above)
Comment (joonhaengHeo) [reply src/controller/java/AndroidCheckInDelegate.cpp:83]: Added. Thanks.

File: src/controller/java/AndroidCheckInDelegate.cpp (hunk 4 above)
Comment (joonhaengHeo) [reply src/controller/java/AndroidCheckInDelegate.cpp:99]: Removed.

File: src/controller/java/AndroidCheckInDelegate.cpp (hunk 5 above)
Comment (joonhaengHeo) [reply src/controller/java/AndroidCheckInDelegate.cpp:80]: Modified. Thanks.

File: src/controller/java/AndroidCheckInDelegate.cpp (hunk 6 above)
Comment (joonhaengHeo) [reply src/controller/java/AndroidCheckInDelegate.cpp:62]: Added.

**Review by yunhanw-google — APPROVED** [review]

**Review by sharadb-amazon — APPROVED** [review]

**Pull request merged**